crypto/bn/bn_exp.c

```diff
@@ -648,34 +648,41 @@ int BN_mod_exp_mont_consttime(BIGNUM *rr, const BIGNUM *a, const BIGNUM *p,
             goto err;
     }
 
+    if (a->neg || BN_ucmp(a, m) >= 0) {
```

Comment:
On related *side* note one can wonder if it makes sense to make BN_ucmp constant time. Specifically for a->top == m->top case. [Or maybe even a->top <= m->top?] Concern is that variation could give away some information about most significant limb[s]. The probability of variation in a->top == m->top case is very low (it's likely that it's sufficient to compare just one limb), but formally speaking the potential is there. In this sense a->top < m->top is actually more "dangerous", because it would give strong signal from observation whether whole of BN_ucmp is executed or just part of it, up to tops' comparison. It might be possible to argue that it's sufficient to ensure that whole BN_ucmp is executed and let unlike variations alone as too unlike and too small to catch... (Just in case for reference, "smooth RSA" won't pass a->top < m->top.)

"Side note" means that no action is expected in the context of this request. By itself, it's ready to roll.

Reply:
Yup. We just made `BN_cmp` constant-time in BoringSSL. There's something to be said for separate constant-time `BN_equals` and `BN_less_than` functions since you can be a hair more efficient when you tell the function which constant-time comparison you're actually interested in, but there are enough `BN_cmp` callers inside and outside the library that we just fixed the original function.

`a->top < m->top` I think should be solved by fixing that case to begin with. Otherwise it's too hard to reason about where things are okay and one needs to expend a lot of complexity to mask off entries past `a->top`.

Reply:
> `a->top < m->top` I think should be solved by fixing that case to begin with.

Yes, of course. Unfortunately there is "modulo" thing:-( Modulo fact that solution won't be back-port-able, and it might still be important to understand and classify problem, this or any other, and recognize if it has to be addressed in maintenance branches and how.

To elaborate on my remark. What does "probability of variation in a->top == m->top" have to do with anything? It's actually likely that single comparison would suffice. But let's say it's the unlike event and it takes two. What would be the variation? Or rather would attacker be able to [amplify and] measure it [reliably enough to mount an attack]? Three comparisons are even less likely and variation is still virtually as hard to measure... Is there anything one can do to minimize the variation in either case? And now connecting back to a->top < m->top case. It's possible to fake single-comparison code path. Would it be appropriate to do that?

Note that this is *not* argument against the suggestion to move toward real/omniscient solution, it's merely an attempt to understand how susceptible maintenance branches would be. With implied question if one can live with that or what can/should one do about it.

On remotely related note regarding bn_correct_top. One probably should recognize that impact from handling "fixed top" is the one that is more problematic than "fixing top" itself. I mean variations from bn_correct_top itself is hardly problem, or at least much-much-much[!] of a lesser problem than variation from taking different code paths depending on uneven tops.

Reply:
It occurred to me that remark can appear somewhat nonsensical in the context of the specific problem this merge request is solving. I mean the request is solving the problem with input being equal to modulus, so that we are looking at **all** a-limbs being equal to m-limbs. While remark refers to probability of already top-most limbs being equal as low. But there is no contradiction, because the reported problem is with **uncommon** inputs, for which we **don't** actually aim to provide constant-time-ness. Constant-time-ness is for **common** inputs and remark refers to these common inputs, for which low-probability assertion holds. Just in case, remote attacker customarily is not in position to have such inputs passed to subroutine in question, which is why lack of constant-time-ness for uncommon inputs in not considered problematic.